src/bitwarden_fmt_plug.c

```diff
@@ -0,0 +1,221 @@
+ */
+
+#if FMT_EXTERNS_H
+extern struct fmt_main fmt_bitwarden;
+#elif FMT_REGISTERS_H
+john_register_one(&fmt_bitwarden);
+#else
+
+#include <string.h>
+
+#ifdef _OPENMP
+#include <omp.h>
+#ifndef OMP_SCALE
+#define OMP_SCALE               8
+#endif
+#endif
+
+#include "arch.h"
+#include "misc.h"
+#include "common.h"
+#include "formats.h"
+#include "params.h"
+#include "options.h"
+#include "jumbo.h"
+#include "aes.h"
+#include "pbkdf2_hmac_sha256.h"
+#include "bitwarden_common.h"
+#include "memdbg.h"
+
+#define FORMAT_LABEL            "Bitwarden"
+#ifdef SIMD_COEF_32
+#define ALGORITHM_NAME          "PBKDF2-SHA256 AES " SHA256_ALGORITHM_NAME
+#else
+#if ARCH_BITS >= 64
+#define ALGORITHM_NAME          "PBKDF2-SHA256 AES 64/" ARCH_BITS_STR " " SHA2_LIB
+#else
+#define ALGORITHM_NAME          "PBKDF2-SHA256 AES 32/" ARCH_BITS_STR " " SHA2_LIB
+#endif
+#endif
+#define BENCHMARK_COMMENT       ""
+#define BENCHMARK_LENGTH        -1
+#define BINARY_SIZE             0
+#define PLAINTEXT_LENGTH        125
+#define SALT_SIZE               sizeof(struct custom_salt)
+#define BINARY_ALIGN            1
+#define SALT_ALIGN              sizeof(uint64_t)
+#ifdef SIMD_COEF_32
+#define MIN_KEYS_PER_CRYPT      SSE_GROUP_SZ_SHA256
+#define MAX_KEYS_PER_CRYPT      SSE_GROUP_SZ_SHA256
+#else
+#define MIN_KEYS_PER_CRYPT      1
+#define MAX_KEYS_PER_CRYPT      1
+#endif
+
+static char (*saved_key)[PLAINTEXT_LENGTH + 1];
+static int *cracked, cracked_count;
+static struct custom_salt *cur_salt;
+
+static void init(struct fmt_main *self)
+{
+#if defined (_OPENMP)
+	omp_autotune(self, OMP_SCALE);
+#endif
+	saved_key = mem_calloc(sizeof(*saved_key),  self->params.max_keys_per_crypt);
+	cracked = mem_calloc(sizeof(*cracked), self->params.max_keys_per_crypt);
+	cracked_count = self->params.max_keys_per_crypt;
+}
+
+static void done(void)
+{
+	MEM_FREE(cracked);
+	MEM_FREE(saved_key);
+}
+
+static void set_salt(void *salt)
+{
+	cur_salt = (struct custom_salt *)salt;
+}
+
+static void bitwarden_set_key(char *key, int index)
+{
+	strnzcpy(saved_key[index], key, sizeof(*saved_key));
+}
+
+static char *get_key(int index)
+{
+	return saved_key[index];
+}
+
+static int bitwarden_decrypt(struct custom_salt *cur_salt, unsigned char *key)
+{
+	unsigned char ivec[IVLEN];
+	unsigned char out[BLOBLEN];
+	AES_KEY aes_decrypt_key;
+
+        AES_set_decrypt_key(key, 256, &aes_decrypt_key);
+        // memcpy(ivec, cur_salt->iv, 16);
+        AES_cbc_encrypt(cur_salt->blob + BLOBLEN - 32, out, BLOBLEN, &aes_decrypt_key, ivec, AES_DECRYPT);
+
+	return memcmp(out + 16, "\x10\x10\x10\x10\x10\x10\x10\x10\x10\x10\x10\x10\x10\x10\x10\x10", 16) == 0;
+}
+
+static int crypt_all(int *pcount, struct db_salt *salt)
+{
+	const int count = *pcount;
+	int index = 0;
+
+	memset(cracked, 0, sizeof(cracked[0]) * cracked_count);
+
+#ifdef _OPENMP
+#pragma omp parallel for
+#endif
+	for (index = 0; index < count; index += MAX_KEYS_PER_CRYPT) {
```

Comment:
All uses of MAX_KEYS_PER_CRYPT within crypt_all here should probably better be MIN_KEYS_PER_CRYPT. Otherwise we can't bump MAX_KEYS_PER_CRYPT.

Reply:
Hmm. I have made this change now but I don't really understand it yet. Maybe it will click "soon" ;)

Reply:
Right now they are the same - and probably for this format that will stay. But once you'd bump MAX_KEYS_PER_CRYPT for performance reasons, it would fail unless you change this. What we really want here is "number of keys per natural block" which is 1 for non-SIMD and 4-24 or even greater for SIMD (with interleaving).